src/npyffi/array.rs

```diff
@@ -31,28 +31,35 @@ pub static PY_ARRAY_API: PyArrayAPI = PyArrayAPI::new();
 
 /// See [PY_ARRAY_API] for more.
 pub struct PyArrayAPI {
-    api: Cell<*const *const c_void>,
+    api: AtomicUsize,
 }
 
 impl PyArrayAPI {
     const fn new() -> Self {
         Self {
-            api: Cell::new(ptr::null_mut()),
+            api: AtomicUsize::new(0),
         }
     }
-    fn get(&self, offset: isize) -> *const *const c_void {
-        if self.api.get().is_null() {
-            Python::with_gil(|py| {
-                let api = get_numpy_api(py, MOD_NAME, CAPSULE_NAME);
-                self.api.set(api);
-            });
+    #[cold]
+    fn init(&self) -> *const *const c_void {
+        Python::with_gil(|py| {
+            let mut api = self.api.load(Ordering::Relaxed) as *const *const c_void;
+            if api.is_null() {
+                api = get_numpy_api(py, MOD_NAME, CAPSULE_NAME);
```

Comment:
As a potential gotcha, can `get_numpy_api` lead to temporary release of the GIL lock? That would potentially enable multiple threads to run this initialization.

Reply:
As CPython's import implementation can be hooked, I think one cannot prevent this from happening in general. But I also think  that multiple threads performing the initialization is only an issue of efficiency.

If a hook is releasing the GIL for whatever reason, it needs to be reacquired and all threads will only progress back here with the GIL held and at most store the same capsule pointer redundantly. (Doing the double-checking here on my part was only motivated by efficiency, i.e. we already have to take the lock so why not use this to avoid redundant initialization as we are already on the slow path.)

(If multiple threads importing the same module yields a different capsule and hence API pointer, I think all bets are off and we would need external synchronization like using `std::sync::Once`.)

Reply:
> (If multiple threads importing the same module yields a different capsule and hence API pointer, I think all bets are off and we would need external synchronization like using std::sync::Once.)

Well we could `compare_exchange` the pointer instead of `store`ing it and only update it if it still NULL and otherwise discard our just initialized value in favour of the "old" returned by `compare_exchange`.

But having the global at all seems weird if we are expecting that the `get_numpy_api` returns different capsules when called from different threads or at different times.

Reply:
Agreed, I think most likely this code is fine as-is thanks to the global nature. In [`GILOnceCell`](https://docs.rs/pyo3/0.15.1/pyo3/once_cell/struct.GILOnceCell.html) I chose to drop any surplus values produced by other threads if a race occurred. This was kind of necessary because of the API contract of it being write-once.